File: agent/tools/policy_delete.go
package tools

import (
	"fmt"

	"github.com/sirupsen/logrus"

	"github.com/vultisig/vultisig-win/agent/signing"
	"github.com/vultisig/vultisig-win/agent/verifier"
	"github.com/vultisig/vultisig-win/tss"
)

const deleteDerivePath = "m/44'/60'/0'/0/0"

type PolicyDeleteTool struct {
	client *verifier.Client
	tss    *tss.TssService
	logger *logrus.Logger
}

func NewPolicyDeleteTool(client *verifier.Client, tss *tss.TssService) *PolicyDeleteTool {
	return &PolicyDeleteTool{
		client: client,
		tss:    tss,
		logger: logrus.WithField("module", "policy_delete").Logger,
	}
}

func (t *PolicyDeleteTool) Name() string {
	return "policy_delete"
}

func (t *PolicyDeleteTool) Description() string {
	return "Delete a policy. This requires TSS signing."
}

func (t *PolicyDeleteTool) InputSchema() map[string]any {
	return map[string]any{
		"policy_id": map[string]any{
			"type":        "string",
			"description": "The ID of the policy to delete",
		},
	}
}

func (t *PolicyDeleteTool) RequiresPassword() bool {
	return true
}

func (t *PolicyDeleteTool) RequiresConfirmation() bool {
	return true
}

func (t *PolicyDeleteTool) Execute(input map[string]any, ctx *ExecutionContext) (any, error) {
	policyIDRaw, ok := input["policy_id"]
	if !ok {
		return nil, fmt.Errorf("policy_id is required")
	}

	policyID, ok := policyIDRaw.(string)
	if !ok {
		return nil, fmt.Errorf("policy_id must be a string")
	}

	policyDetails, err := t.client.GetPolicy(policyID, ctx.AuthToken)
	if err != nil {
		return nil, fmt.Errorf("failed to fetch policy: %w", err)
	}

	pluginVersion := policyDetails.PluginVersion
	if pluginVersion == "" {
		pluginVersion = "1.0.0"
	}

	messageToSign := fmt.Sprintf(
		"%s*#*%s*#*%d*#*%s",
		policyDetails.Recipe,
		policyDetails.PublicKey,
		policyDetails.PolicyVersion,
		pluginVersion,
	)

	t.logger.WithField("message", messageToSign).Debug("Signing delete message")

	messageHash := signing.EthereumSignHash(messageToSign)
	cfg := KeysignConfig{
		AppCtx:     ctx.AppCtx,
		Ctx:        ctx.Ctx,
		Vault:      ctx.Vault,
		Password:   ctx.Password,
		DerivePath: deleteDerivePath,
		TSS:        t.tss,
		Logger:     t.logger,
	}
	signature, err := FastVaultKeysign(cfg, messageHash)
	if err != nil {
		return nil, fmt.Errorf("failed to sign delete request: %w", err)
	}

	t.logger.Info("Delete signed successfully")

	err = t.client.DeletePolicy(policyID, signature, ctx.AuthToken)
	if err != nil {
		return nil, fmt.Errorf("failed to delete policy: %w", err)
	}

	return map[string]any{
		"success":   true,
		"policy_id": policyID,
		"message":   fmt.Sprintf("Policy %s deleted successfully.", policyID),
		"ui": map[string]any{
			"title": "Policy Deleted",
			"actions": []map[string]any{
				{
					"type":  "copy",
					"label": "Copy Policy ID",
					"value": policyID,
				},
			},
		},
	}, nil
}

File: core/ui/agent/tools/handlers/policyDelete.ts
import { ethereumSignHash } from '../shared/ethereumSigning'
import {
  fastVaultKeysign,
  formatKeysignSignatureHex,
} from '../shared/fastVaultKeysign'
import { deletePolicy, getPolicyFull } from '../shared/verifierApi'
import type { ToolHandler } from '../types'

const deleteDerivePath = "m/44'/60'/0'/0/0"

export const handlePolicyDelete: ToolHandler = async (input, context) => {
  const vault = context.vault
  if (!vault) {
    throw new Error('Vault metadata required for policy deletion')
  }

  const policyId = input.policy_id as string
  if (!policyId) throw new Error('policy_id is required')

  const authToken = context.authToken ?? ''

  const policyDetails = await getPolicyFull(policyId, authToken)

  const pluginVersion =
    ((policyDetails as Record<string, unknown>).plugin_version as string) ||
    '1.0.0'

  const policyVersion =
    ((policyDetails as Record<string, unknown>).policy_version as number) ?? 0

  const recipe = policyDetails.recipe ?? ''

  const messageToSign = `${recipe}*#*${policyDetails.public_key}*#*${policyVersion}*#*${pluginVersion}`

  const messageHash = ethereumSignHash(messageToSign)
  const keysignResult = await fastVaultKeysign({
    vault,
    messageHash,
    derivePath: deleteDerivePath,
  })
  const signature = formatKeysignSignatureHex(keysignResult)

  await deletePolicy(policyId, signature, authToken)

  return {
    data: {
      success: true,
      policy_id: policyId,
      message: `Policy ${policyId} deleted successfully.`,
      ui: {
        title: 'Policy Deleted',
        actions: [{ type: 'copy', label: 'Copy Policy ID', value: policyId }],
      },
    },
  }
}
